File: modules/songseed-pitch-shift/android/src/main/java/expo/modules/songseedpitchshift/SongseedPitchShiftPracticeEngine.kt
package expo.modules.songseedpitchshift

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.PlaybackParameters
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer
import java.util.concurrent.CountDownLatch
import kotlin.math.max
import kotlin.math.min

private const val DEFAULT_TICK_MS = 100L

class SongseedPitchShiftPracticeEngine(
  context: Context,
  private val onStateChange: (Map<String, Any?>) -> Unit,
  private val onPlaybackEnded: (Map<String, Any?>) -> Unit,
  private val onError: (String) -> Unit,
) {
  private val appContext = context.applicationContext
  private val mainHandler = Handler(Looper.getMainLooper())
  private var player: ExoPlayer? = null
  private var currentSourceUri: String? = null
  private var currentPlaybackRate = 1.0
  private var currentPitchShiftSemitones = 0
  private var didJustFinish = false
  private var tickerRunning = false

  private val ticker = object : Runnable {
    override fun run() {
      if (!tickerRunning) {
        return
      }
      onStateChange(getStateUnsafe())
      mainHandler.postDelayed(this, DEFAULT_TICK_MS)
    }
  }

  private fun <T> runOnPlayerThreadSync(action: () -> T): T {
    if (Looper.myLooper() == Looper.getMainLooper()) {
      return action()
    }

    val latch = CountDownLatch(1)
    var result: Any? = null
    var error: Throwable? = null

    mainHandler.post {
      try {
        result = action()
      } catch (throwable: Throwable) {
        error = throwable
      } finally {
        latch.countDown()
      }
    }

    latch.await()
    error?.let { throw it }

    @Suppress("UNCHECKED_CAST")
    return result as T
  }

  private fun ensurePlayer(): ExoPlayer {
    player?.let { return it }

    val created = ExoPlayer.Builder(appContext).build().apply {
      setAudioAttributes(
        AudioAttributes.Builder()
          .setUsage(C.USAGE_MEDIA)
          .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
          .build(),
        true
      )
          addListener(
            object : Player.Listener {
              override fun onPlaybackStateChanged(playbackState: Int) {
                if (playbackState == Player.STATE_ENDED) {
                  didJustFinish = true
                  val state = getStateUnsafe()
                  onPlaybackEnded(state)
                  onStateChange(state)
                } else {
                  onStateChange(getStateUnsafe())
                }
              }

              override fun onIsPlayingChanged(isPlaying: Boolean) {
                if (isPlaying) {
              startTicker()
                } else if ((player?.playbackState ?: Player.STATE_IDLE) == Player.STATE_ENDED) {
                  stopTicker()
                }
                onStateChange(getStateUnsafe())
              }

              override fun onMediaItemTransition(mediaItem: MediaItem?, reason: Int) {
                didJustFinish = false
                onStateChange(getStateUnsafe())
              }

              override fun onPlayerError(error: PlaybackException) {
                stopTicker()
                didJustFinish = false
                onError(error.message ?: "Android pitch shift playback failed.")
                onStateChange(getStateUnsafe())
              }
            }
          )
    }

    player = created
    return created
  }

  private fun startTicker() {
    if (tickerRunning) return
    tickerRunning = true
    mainHandler.post(ticker)
  }

  private fun stopTicker() {
    tickerRunning = false
    mainHandler.removeCallbacks(ticker)
  }

  private fun clampPlaybackRate(value: Double): Float {
    return max(0.5f, min(2.0f, value.toFloat()))
  }

  private fun applyPlaybackParameters(target: ExoPlayer) {
    val pitchMultiplier = Math.pow(2.0, currentPitchShiftSemitones / 12.0).toFloat()
    target.playbackParameters = PlaybackParameters(clampPlaybackRate(currentPlaybackRate), pitchMultiplier)
  }

  fun getCapabilities(): Map<String, Any> {
    return mapOf(
      "isAvailable" to true,
      "supportsPracticePlayback" to true,
      "supportsEditorPreview" to false,
      "supportsOfflineRender" to false,
      "minSemitones" to -12,
      "maxSemitones" to 12,
    )
  }

  private fun getStateUnsafe(): Map<String, Any?> {
    val target = player
    val durationMs =
      when {
        target == null -> 0
        target.duration == C.TIME_UNSET -> 0
        else -> target.duration
      }

    return mapOf(
      "isAvailable" to true,
      "isLoaded" to (target != null && currentSourceUri != null),
      "isPlaying" to (target?.isPlaying ?: false),
      "didJustFinish" to didJustFinish,
      "currentTimeMs" to (target?.currentPosition ?: 0L),
      "durationMs" to durationMs,
      "playbackRate" to currentPlaybackRate,
      "pitchShiftSemitones" to currentPitchShiftSemitones,
      "sourceUri" to currentSourceUri,
      )
  }

  fun getState(): Map<String, Any?> = runOnPlayerThreadSync {
    getStateUnsafe()
  }

  fun loadForPractice(request: Map<String, Any?>): Map<String, Any?> {
    return runOnPlayerThreadSync {
      val sourceUri = request["sourceUri"] as? String
        ?: throw IllegalArgumentException("Pitch shift practice playback requires sourceUri.")
      val startPositionMs = (request["startPositionMs"] as? Number)?.toLong() ?: 0L
      val autoplay = request["autoplay"] as? Boolean ?: false
      currentPlaybackRate = (request["playbackRate"] as? Number)?.toDouble() ?: 1.0
      currentPitchShiftSemitones = (request["pitchShiftSemitones"] as? Number)?.toInt() ?: 0
      didJustFinish = false
      currentSourceUri = sourceUri

      val target = ensurePlayer()
      stopTicker()
      target.stop()
      target.clearMediaItems()
      target.setMediaItem(MediaItem.fromUri(Uri.parse(sourceUri)))
      applyPlaybackParameters(target)
      target.prepare()
      target.seekTo(max(0L, startPositionMs))
      target.playWhenReady = autoplay
      if (autoplay) {
        startTicker()
      } else {
        onStateChange(getStateUnsafe())
      }
      getStateUnsafe()
    }
  }

  fun play(): Map<String, Any?> = runOnPlayerThreadSync {
    val target = player ?: return@runOnPlayerThreadSync getStateUnsafe()
    val durationMs =
      when {
        target.duration == C.TIME_UNSET -> 0L
        else -> target.duration
      }
    val isAtEnd = durationMs > 0L && target.currentPosition >= max(0L, durationMs - 50L)
    if (didJustFinish || target.playbackState == Player.STATE_ENDED || isAtEnd) {
      target.seekTo(0)
    }
    didJustFinish = false
    target.playWhenReady = true
    target.play()
    startTicker()
    getStateUnsafe()
  }

  fun pause(): Map<String, Any?> = runOnPlayerThreadSync {
    val target = player ?: return@runOnPlayerThreadSync getStateUnsafe()
    target.pause()
    stopTicker()
    getStateUnsafe()
  }

  fun stop(): Map<String, Any?> = runOnPlayerThreadSync {
    val target = player ?: return@runOnPlayerThreadSync getStateUnsafe()
    didJustFinish = false
    target.pause()
    target.seekTo(0)
    stopTicker()
    getStateUnsafe()
  }

  fun unload(): Map<String, Any?> = runOnPlayerThreadSync {
    stopTicker()
    didJustFinish = false
    player?.release()
    player = null
    currentSourceUri = null
    currentPlaybackRate = 1.0
    currentPitchShiftSemitones = 0
    getStateUnsafe()
  }

  fun seekTo(positionMs: Double): Map<String, Any?> = runOnPlayerThreadSync {
    val target = player ?: return@runOnPlayerThreadSync getStateUnsafe()
    didJustFinish = false
    target.seekTo(max(0L, positionMs.toLong()))
    getStateUnsafe()
  }

  fun setPlaybackRate(rate: Double): Map<String, Any?> = runOnPlayerThreadSync {
    currentPlaybackRate = rate
    player?.let { applyPlaybackParameters(it) }
    getStateUnsafe()
  }

  fun setPitchShiftSemitones(semitones: Int): Map<String, Any?> = runOnPlayerThreadSync {
    currentPitchShiftSemitones = semitones.coerceIn(-12, 12)
    didJustFinish = false
    player?.let { applyPlaybackParameters(it) }
    getStateUnsafe()
  }
}

File: modules/songseed-pitch-shift/ios/SongseedPitchShiftPracticeEngine.swift
import AVFoundation

private let pitchShiftTickInterval: TimeInterval = 0.1

final class SongseedPitchShiftPracticeEngine {
  private let onStateChange: ([String: Any]) -> Void
  private let onPlaybackEnded: ([String: Any]) -> Void
  private let onError: (String) -> Void

  private let engine = AVAudioEngine()
  private let playerNode = AVAudioPlayerNode()
  private let timePitchNode = AVAudioUnitTimePitch()

  private var audioFile: AVAudioFile?
  private var sourceUri: String?
  private var currentPlaybackRate: Double = 1.0
  private var currentPitchShiftSemitones: Int = 0
  private var pausedPositionMs: Double = 0
  private var isPlaying = false
  private var didJustFinish = false
  private var playbackSessionId = UUID()
  private var tickTimer: Timer?

  init(
    onStateChange: @escaping ([String: Any]) -> Void,
    onPlaybackEnded: @escaping ([String: Any]) -> Void,
    onError: @escaping (String) -> Void
  ) {
    self.onStateChange = onStateChange
    self.onPlaybackEnded = onPlaybackEnded
    self.onError = onError

    engine.attach(playerNode)
    engine.attach(timePitchNode)
    engine.connect(playerNode, to: timePitchNode, format: nil)
    engine.connect(timePitchNode, to: engine.mainMixerNode, format: nil)
    applyTimePitchSettings()
  }

  deinit {
    tickTimer?.invalidate()
    playerNode.stop()
    engine.stop()
  }

  func getCapabilities() -> [String: Any] {
    [
      "isAvailable": true,
      "supportsPracticePlayback": true,
      "supportsEditorPreview": true,
      "supportsOfflineRender": true,
      "minSemitones": -12,
      "maxSemitones": 12
    ]
  }

  func getState() -> [String: Any] {
    [
      "isAvailable": true,
      "isLoaded": audioFile != nil && sourceUri != nil,
      "isPlaying": isPlaying,
      "didJustFinish": didJustFinish,
      "currentTimeMs": Int(computedCurrentPositionMs().rounded()),
      "durationMs": durationMs(),
      "playbackRate": currentPlaybackRate,
      "pitchShiftSemitones": currentPitchShiftSemitones,
      "sourceUri": sourceUri as Any
    ]
  }

  func loadForPractice(_ request: [String: Any]) throws -> [String: Any] {
    guard let uriString = request["sourceUri"] as? String else {
      throw NSError(
        domain: "SongseedPitchShift",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "Pitch shift practice playback requires sourceUri."]
      )
    }

    let startPositionMs = (request["startPositionMs"] as? Double)
      ?? Double((request["startPositionMs"] as? Int) ?? 0)
    let autoplay = request["autoplay"] as? Bool ?? false
    let playbackRate = (request["playbackRate"] as? Double)
      ?? Double((request["playbackRate"] as? Int) ?? 1)
    let semitones = (request["pitchShiftSemitones"] as? Int)
      ?? Int((request["pitchShiftSemitones"] as? Double) ?? 0)

    let url = URL(string: uriString) ?? URL(fileURLWithPath: uriString)
    let file = try AVAudioFile(forReading: url)

    stopTimer()
    playerNode.stop()
    audioFile = file
    sourceUri = uriString
    currentPlaybackRate = max(0.5, min(2.0, playbackRate))
    currentPitchShiftSemitones = max(-12, min(12, semitones))
    pausedPositionMs = max(0, startPositionMs)
    didJustFinish = false
    isPlaying = false
    applyTimePitchSettings()
    try ensureEngineRunning()

    if autoplay {
      try playInternal()
    } else {
      onStateChange(getState())
    }

    return getState()
  }

  func play() throws -> [String: Any] {
    let duration = Double(durationMs())
    if didJustFinish || (duration > 0 && pausedPositionMs >= max(0, duration - 50)) {
      pausedPositionMs = 0
    }
    try playInternal()
    return getState()
  }

  func pause() -> [String: Any] {
    pausedPositionMs = computedCurrentPositionMs()
    playerNode.stop()
    isPlaying = false
    stopTimer()
    onStateChange(getState())
    return getState()
  }

  func stop() -> [String: Any] {
    pausedPositionMs = 0
    didJustFinish = false
    playerNode.stop()
    isPlaying = false
    stopTimer()
    onStateChange(getState())
    return getState()
  }

  func unload() -> [String: Any] {
    stopTimer()
    playerNode.stop()
    engine.stop()
    audioFile = nil
    sourceUri = nil
    pausedPositionMs = 0
    currentPlaybackRate = 1.0
    currentPitchShiftSemitones = 0
    isPlaying = false
    didJustFinish = false
    applyTimePitchSettings()
    return getState()
  }

  func seekTo(_ positionMs: Double) throws -> [String: Any] {
    pausedPositionMs = max(0, min(positionMs, Double(durationMs())))
    didJustFinish = false

    if isPlaying {
      try playInternal()
    } else {
      onStateChange(getState())
    }

    return getState()
  }

  func setPlaybackRate(_ rate: Double) -> [String: Any] {
    currentPlaybackRate = max(0.5, min(2.0, rate))
    didJustFinish = false
    applyTimePitchSettings()
    onStateChange(getState())
    return getState()
  }

  func setPitchShiftSemitones(_ semitones: Int) -> [String: Any] {
    currentPitchShiftSemitones = max(-12, min(12, semitones))
    didJustFinish = false
    applyTimePitchSettings()
    onStateChange(getState())
    return getState()
  }

  private func applyTimePitchSettings() {
    timePitchNode.rate = Float(currentPlaybackRate)
    timePitchNode.pitch = Float(currentPitchShiftSemitones * 100)
  }

  private func ensureEngineRunning() throws {
    if engine.isRunning {
      return
    }

    do {
      try engine.start()
    } catch {
      onError(error.localizedDescription)
      throw error
    }
  }

  private func playInternal() throws {
    guard audioFile != nil else {
      return
    }

    try ensureEngineRunning()
    didJustFinish = false
    playbackSessionId = UUID()
    scheduleFromCurrentPosition(sessionId: playbackSessionId)
    playerNode.play()
    isPlaying = true
    startTimer()
    onStateChange(getState())
  }

  private func scheduleFromCurrentPosition(sessionId: UUID) {
    guard let file = audioFile else {
      return
    }

    playerNode.stop()

    let sampleRate = file.processingFormat.sampleRate
    let totalFrames = file.length
    let startFrame = AVAudioFramePosition((pausedPositionMs / 1000.0) * sampleRate)
    let safeStartFrame = max(0, min(startFrame, totalFrames))
    let remainingFrames = max(0, totalFrames - safeStartFrame)

    if remainingFrames <= 0 {
      pausedPositionMs = Double(durationMs())
      isPlaying = false
      stopTimer()
      onStateChange(getState())
      return
    }

    playerNode.scheduleSegment(
      file,
      startingFrame: safeStartFrame,
      frameCount: AVAudioFrameCount(remainingFrames),
      at: nil
    ) { [weak self] in
      DispatchQueue.main.async {
        guard let self, self.playbackSessionId == sessionId else {
          return
        }
        self.pausedPositionMs = Double(self.durationMs())
        self.isPlaying = false
        self.didJustFinish = true
        self.stopTimer()
        let state = self.getState()
        self.onPlaybackEnded(state)
        self.onStateChange(state)
      }
    }
  }

  private func computedCurrentPositionMs() -> Double {
    guard
      let file = audioFile,
      isPlaying,
      let nodeTime = playerNode.lastRenderTime,
      let playerTime = playerNode.playerTime(forNodeTime: nodeTime)
    else {
      return pausedPositionMs
    }

    let sampleRate = file.processingFormat.sampleRate
    let renderedMs = (Double(playerTime.sampleTime) / sampleRate) * 1000.0
    return max(0, min(pausedPositionMs + renderedMs, Double(durationMs())))
  }

  private func durationMs() -> Int {
    guard let file = audioFile else {
      return 0
    }
    let durationSeconds = Double(file.length) / file.processingFormat.sampleRate
    return Int((durationSeconds * 1000.0).rounded())
  }

  private func startTimer() {
    stopTimer()
    tickTimer = Timer.scheduledTimer(withTimeInterval: pitchShiftTickInterval, repeats: true) { [weak self] _ in
      guard let self else {
        return
      }
      self.onStateChange(self.getState())
    }
    RunLoop.main.add(tickTimer!, forMode: .common)
  }

  private func stopTimer() {
    tickTimer?.invalidate()
    tickTimer = nil
  }
}
